src/int128_struct_impl.h

```diff
@@ -0,0 +1,137 @@
+#ifndef SECP256K1_INT128_STRUCT_IMPL_H
+#define SECP256K1_INT128_STRUCT_IMPL_H
+
+#include "int128.h"
+
+#if defined(_MSC_VER) && (defined(_M_X64) || defined(_M_ARM64)) /* MSVC */
+  #include <intrin.h>
+  #define secp256k1_umulh __umulh
+  #define secp256k1_mulh __mulh
+#else
+static SECP256K1_INLINE uint64_t secp256k1_umulh(uint64_t a, uint64_t b) {
+   uint64_t t1 = (uint64_t)(uint32_t)a * (uint32_t)b;
+   uint64_t t2 = (a >> 32) * (uint32_t)b;
+   uint64_t t3 = (uint32_t)a*(b >> 32) + (t1 >> 32) + (uint32_t)t2;
+   return (a >> 32)*(b >> 32) + (t2 >> 32) + (t3 >> 32);
+}
+
+static SECP256K1_INLINE int64_t secp256k1_mulh(int64_t a, int64_t b) {
+   uint64_t t1 = (uint64_t)(uint32_t)a * (uint32_t)b;
+   int64_t t2 = (a >> 32) * (uint32_t)b;
+   int64_t t3 = (uint32_t)a * (b >> 32);
+   uint64_t t4 = (t1 >> 32) + (uint32_t)t2 + (uint32_t)t3;
+   return (a >> 32) * (b >> 32) + (t2 >> 32) + (t3 >> 32) + (int64_t)(t4 >> 32);
+}
+#endif
+
+static SECP256K1_INLINE void secp256k1_u128_mul(secp256k1_uint128 *r, uint64_t a, uint64_t b) {
+   r->hi = secp256k1_umulh(a, b);
+   r->lo = a * b;
+}
+
+static SECP256K1_INLINE void secp256k1_u128_accum_mul(secp256k1_uint128 *r, uint64_t a, uint64_t b) {
+   uint64_t lo = a * b;
+   r->hi += secp256k1_umulh(a, b) + (~lo < r->lo);
+   r->lo += lo;
+}
+
+static SECP256K1_INLINE void secp256k1_u128_accum_u64(secp256k1_uint128 *r, uint64_t a) {
+   r->hi += (r->lo > ~a);
+   r->lo += a;
+}
+
+/* Unsigned (logical) right shift.
+ * Non-constant time in n.
+ */
+static SECP256K1_INLINE void secp256k1_u128_rshift(secp256k1_uint128 *r, unsigned int n) {
+   VERIFY_CHECK(n < 128);
+   if (n >= 64) {
+     r->lo = (r->hi) >> (n-64);
+     r->hi = 0;
+   } else if (n > 0) {
+     r->lo = ((r->hi & (((uint64_t)1<<n)-1)) << (64-n)) | r->lo >> n;
```

Comment:
@real-or-random Unsigned overflow from shifting is well defined right?  So I don't need the mask here?

Reply:
Right, unsigned overflow is never UB, even when it comes from shifting. See C99: https://port70.net/~nsz/c/c99/n1256.html#6.5.7p4 

(The corresponding C89 part is a little bit strange since it doesn't mention the UB explicitly and it works under the assumption that `int` and `long` are the only non-promoted integer types: https://port70.net/~nsz/c/c89/c89-draft.html#3.3.7 I mean ok, `stdint.h` does not exist in C89...)

Reply:
Oh I remember why I always write it this way, because we _technically_ don't know the size of a signed integer.

Reply:
yes, unsigned overflow is always defined, it's wrap-around no matter what

Reply:
Right.  But *technically speaking* a `signed int` could, in some unrealistic universe, be 65 bits, thus subjecting `r->hi` to signed integer promotion and thus causing signed integer overflow.

That said, if people would like me to rewrite this to assume 32-bit `signed int`s, I'm happy to change it.

Reply:
Done something similar to the suggestion.